### PR TITLE
[dumpglob] Document the [.glob] file format.

### LINKED ISSUE #4834: document the *.glob file format
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4834
From: @siegebell
Reported version: unspecified
CC:   coq-bugs-redist@lists.gforge.inria.fr

### PR DESCRIPTION
**Pull request by ejgallego**
Fixes #4834

### CODE REVIEW

File: interp/dumpglob.mli [hunk 1]
```diff
@@ -8,6 +8,61 @@
...
+
+[DIGEST: %md5sum_of_file]
+[%modpath]
+
+  where %modpath is the fully-qualified module name of the library that the
+  [.glob] file refers to.
+
+ {3 [.glob] entries }
+
+  The are 2 kinds of [.glob] entries:
```
Comment (herbelin) [inline interp/dumpglob.mli:44]:
```suggestion
  There are 2 kinds of [.glob] entries:
```

File: interp/dumpglob.mli [hunk 2]
```diff
@@ -8,6 +8,61 @@
...
+
+  - *definitions*: these entries correspond to definitions of inductives,
+    functions, binders, notations. They are written as:
+
+    [%type %bc:%ec %secpath %name]
+
+    where type is one of
+    [{def,coe,subclass,canonstruc,ex,def,def,scheme,proj,coe,inst,meth,defax,prfax,prfax,prfax,thm,prim,class,var,indrec,rec,corec,ind,variant,coind,constr,not,binder}]
+    [%bc] and [%ec] are respectively the start and end character for the file, [%secpath] the section path
+    (if any) and [%name] the name of the defined object.
```
Comment (herbelin) [inline interp/dumpglob.mli:54]:
```suggestion
    (or [<>] if none) and [%name] the name of the defined object.
```

**Review by herbelin — APPROVED** [review]
Good initiative!

It does not have to be parsed a priori but if users of the documentation find it helpful, it can be added that notations are encoded under a format `:entry:scope:notation_string` where `_` replaces spaces in the notation string, and where entry is left empty if in constr and scope is also left empty if no scope is attached to the notation.

Also, for binding variables, the added `:number` is to distinguish uniquely in a file the different binding variables of same name.

File: interp/dumpglob.mli [hunk 3]
```diff
@@ -8,6 +8,61 @@
...
+
+  - *references*: which identify the object a particular document piece of text points to;
+    their format is:
+
+    [R%%bc:%ec %filepath %secpath %name %kind]
```
[the remark this answers is not in the document]
Comment (xavierleroy) [reply interp/dumpglob.mli:59]: Also, `name` can be `<>` if the reference is of the `lib` kind.  (And maybe in other cases too.)

**Review by xavierleroy — COMMENTED** [review]
Thanks, Emilio, for your valiant attempt at documenting glob files.  Some comments below.  Globally (pun intended!) you stop just short of the most delicate aspects of the format.

File: interp/dumpglob.mli [hunk 4]
```diff
@@ -8,6 +8,95 @@
...
+
+   [.glob] files contain a header, and then a list of entries, with
+  one line per entry.
+
+  {3 [.glob] header }
+
+  The header consists of two lines:
+
+[DIGEST: %md5sum_of_file]
+[%modpath]
```
Comment (JasonGross) [inline interp/dumpglob.mli:37]: Shouldn't this be
```suggestion
[F%modpath]
```
?  Also, what does F stand for here?

File: interp/dumpglob.mli [hunk 5]
```diff
@@ -8,6 +8,61 @@
...
+  The are 2 kinds of [.glob] entries:
+
+  - *definitions*: these entries correspond to definitions of inductives,
+    functions, binders, notations. They are written as:
+
+    [%type %bc:%ec %secpath %name]
+
+    where type is one of
+    [{def,coe,subclass,canonstruc,ex,def,def,scheme,proj,coe,inst,meth,defax,prfax,prfax,prfax,thm,prim,class,var,indrec,rec,corec,ind,variant,coind,constr,not,binder}]
+    [%bc] and [%ec] are respectively the start and end character for the file, [%secpath] the section path
```
[the remark this answers is not in the document]
Comment (JasonGross) [reply interp/dumpglob.mli:79]: Section paths are explained a couple lines down (not sure if this is new after your comment)

File: interp/dumpglob.mli [hunk 6]
```diff
@@ -8,6 +8,95 @@
...
+    + [not] Notation
+    + [binder] Binder
+
+    [%bc] and [%ec] are respectively the start and end character for the file, [%secpath] the section path
+    (or [<>] if no section path) and [%name] the name of the defined object, or also [<>] in where no name applies.
+
+    Section paths are ...
+
+    + In the case of notations, [%name] is encoded as [:entry:scope:notation_string] where [_] is used to replace
+      spaces in the notation key, [%entry] is left empty if the notation entry is [constr],
```
Comment (JasonGross) [inline interp/dumpglob.mli:85]:
```suggestion
      spaces in the notation string, [%entry] is left empty if the notation entry is [constr],
```
Or else change "string" to "key" on the line above.  Consistency is important

File: interp/dumpglob.mli [hunk 7]
```diff
@@ -8,6 +8,95 @@
...
+
+    + For binding variables, [:number] is added to distinguish uniquely different binding variables of the same name in a file.
+
+  - *references*: which identify the object a particular document piece of text points to;
+    their format is:
+
+    [R%%bc:%ec %filepath %secpath %name %kind]
+
+    where [%bc], [%ec], [%name], and [%kind] are as the above; [%filepath] contains the
+    absolute module path the object that is references lives in.
```
Comment (JasonGross) [inline interp/dumpglob.mli:96]: This is a bit misleading.  It's not the absolute module path, but only the dirpath part of it, i.e., the absolute module path of the file the reference lives in.  (Also "that is references lives" is incorrect grammar here.). The rest of the module path (any non-file non-directory modules) are in the prefix of name.

File: interp/dumpglob.mli [hunk 8]
```diff
@@ -8,6 +8,95 @@
...
+(** This file implements the Coq's [.glob] file format, which
+   provides information about the objects that are defined and referenced
+   from a Coq file.
+
+   The [.glob] file format is notably used by [coqdoc] and [coq2html]
+   to generate links and other documentation meta-data.
+
+   Note that we consider this format a legacy one, and no stability
+   guarantees are provided as of today, as we search to replace this
+   format with a more structured and strongly-typed API.
```
Comment (jfehrle) [inline interp/dumpglob.mli:20]: FWIW, I've been thinking a bit about adding a "jump to definition" feature in CoqIDE (and available to other IDEs), perhaps something I'd look at later this year after further debugger work.  I was thinking in terms of a new call in the IDE protocol (passing filename and offset where the user clicked, returning the location to jump to), which could distinguish references to symbols with the same name in parsed files.

Also maybe a "search by name" feature that takes a substring or regex with dynamic updates of the result for each character the user types.

While I don't quite understand exactly how coqdoc uses the .glob info, these seem closely related.  Are you planning to create the "more structured and strongly-typed API" anytime soon?

File: interp/dumpglob.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (JasonGross) [reply interp/dumpglob.mli:93]: > `kind` was called `type` above.

This seems fixed?  Although it's missing the "lib" option


> Also, `name` can be `<>` if the reference is of the `lib` kind. (And maybe in other cases too.)

Also now mentioned above

File: interp/dumpglob.mli [hunk 9]
```diff
@@ -8,6 +8,95 @@
...
+
+  There are 2 kinds of [.glob] entries:
+
+  - *definitions*: these entries correspond to definitions of inductives,
+    functions, binders, notations. They are written as:
+
+    [%kind %bc:%ec %secpath %name]
+
+    where [%kind] is one of
+    [{def,coe,subclass,canonstruc,ex,scheme,proj,inst,meth,defax,prfax,thm,prim,class,var,indrec,rec,corec,ind,variant,coind,constr,not,binder}],
```
Comment (JasonGross) [inline interp/dumpglob.mli:52]: What about lib and mod (is there also modtype?)

File: interp/dumpglob.mli (hunk 4 above)
Comment (JasonGross) [reply interp/dumpglob.mli:37]: I guess F stands for File?

File: interp/dumpglob.mli (hunk 8 above)
Comment (JasonGross) [reply interp/dumpglob.mli:20]: The relevant bit of glob files for your feature is that coqdoc will look for `R`*start*`:`*end* lines that overlap with the byte range of the identifier coqdoc is considering linking, use the kind part to know how to highlight the identifier, and then use the modpath to determine the file the reference lives in end the name and secpath to determine which identifier in that file to link to.  The anchors are created from the definition sort of lines, which say that a given byte range defines a given kind of thing with a given name and secpath.

If you want "search by name", I think `Search` with a string already does that; `Locate` will then give you the fully qualified name of whatever you're looking for, which can then be used to determine what file and what identifier in that file to jump to.

Comment (jfehrle) [reply interp/dumpglob.mli:20]: IIUC, coqdoc generates html files (such as for the standard library) and the .glob files, but I don't see how the HTML uses the .glob files at all--it just creates HTML tags like the following (in Coq.Init.Datatypes.html).  What am I missing here?

image

> If you want "search by name", I think `Search` with a string already does that; `Locate` will then give you the fully qualified name of whatever you're looking for, which can then be used to determine what file and what identifier in that file to jump to.

Perhaps.  But it's not very convenient for users to have to add two commands to their script and then manually navigate to the appropriate file and line number, then remove the two commands.  Much easier to use as a GUI feature.

Comment (JasonGross) [reply interp/dumpglob.mli:20]: > IIUC, coqdoc generates html files (such as for the standard library) and the .glob files

coqc generates the .glob files (and also .vo files, by consuming .v files).  coqdoc consumes .glob files (combined with the corresponding .v file) to generate .html files.

Comment (JasonGross) [reply interp/dumpglob.mli:20]: > But it's not very convenient for users to have to add two commands to their script and then manually navigate to the appropriate file and line number, then remove the two commands. Much easier to use as a GUI feature.

Right, I'm suggesting that the GUI could use the API that already exists for `Search` and `Locate` to perform this operation for the user.

Comment (ejgallego) [reply interp/dumpglob.mli:20]: The XML protocol actually did have a way to forward the Glob info, it survived by chance tho :)

File: interp/dumpglob.mli (hunk 4 above)
Comment (ejgallego) [reply interp/dumpglob.mli:37]: Yes, thanks!

File: interp/dumpglob.mli [hunk 10]
```diff
@@ -8,6 +8,95 @@
...
+    + [rec] Inductive  (variant)
+    + [corec] Coinductive
+    + [ind] Record
+    + [variant] Record (variant)
+    + [coind] Coinductive Record
+    + [constr] Constructor
+    + [not] Notation
+    + [binder] Binder
+
+    [%bc] and [%ec] are respectively the start and end character for the file, [%secpath] the section path
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply interp/dumpglob.mli:79]: Fixed thanks!

File: interp/dumpglob.mli (hunk 6 above)
Comment (ejgallego) [reply interp/dumpglob.mli:85]: changed to key as indeed notations are internally indexed by this string, thanks

**Comment by ejgallego** [comment]
New pass done, I suggest we merge this PR as a checkpoint.

File: interp/dumpglob.mli (hunk 9 above)
Comment (SkySkimmer) [reply interp/dumpglob.mli:52]: I guess
https://github.com/coq/coq/blob/d9d51353214eb04dea89285b1a8713a92b9d82b2/vernac/vernacentries.ml#L1461
several "mod" and "modtype" in the same file

Comment (ejgallego) [reply interp/dumpglob.mli:52]: Thanks, added as best as I could.

File: interp/dumpglob.mli [hunk 11]
```diff
@@ -8,6 +8,100 @@
...
+
+   [.glob] files contain a header, and then a list of entries, with
+  one line per entry.
+
+  {3 [.glob] header }
+
+  The header consists of two lines:
+
+[DIGEST: %md5sum_of_file]
+[%Fmodpath]
```
Comment (JasonGross) [inline interp/dumpglob.mli:37]:
```suggestion
[F%modpath]
```
The `F` is a literal.

File: interp/dumpglob.mli [hunk 12]
```diff
@@ -8,6 +8,100 @@
...
+
+  There are 2 kinds of [.glob] entries:
+
+  - *definitions*: these entries correspond to definitions of inductives,
+    functions, binders, notations. They are written as:
+
+    [%kind %bc:%ec %secpath %name]
+
+    where [%kind] is one of
+    [{def,coe,subclass,canonstruc,ex,scheme,proj,inst,meth,defax,prfax,thm,prim,class,var,indrec,rec,corec,ind,variant,coind,constr,not,binder}],
```
Comment (JasonGross) [inline interp/dumpglob.mli:52]:
```suggestion
    [{def,coe,subclass,canonstruc,ex,scheme,proj,inst,meth,defax,prfax,thm,prim,class,var,indrec,rec,corec,ind,variant,coind,constr,not,binder,lib,mod,modtype}],
```
Should these be sorted or autogenerated or something?

File: interp/dumpglob.mli [hunk 13]
```diff
@@ -8,6 +8,100 @@
...
+    + In the case of notations, [%name] is encoded as [:entry:scope:notation_key] where [_] is used to replace
+      spaces in the notation key, [%entry] is left empty if the notation entry is [constr],
+      and similarly [%scope] is empty if the corresponding notation has no associated scope.
+
+    + For binding variables, [:number] is added to distinguish uniquely different binding variables of the same name in a file.
+
+  - *references*: which identify the object a particular document piece of text points to;
+    their format is:
+
+    [R%%bc:%ec %filepath %secpath %name %kind]
```
Comment (JasonGross) [inline interp/dumpglob.mli:97]: Do you mean
```suggestion
    [R%bc:%ec %filepath %secpath %name %kind]
```
? (please check, I don't have all this loaded in my head right now)

File: interp/dumpglob.mli [hunk 14]
```diff
@@ -8,6 +8,100 @@
...
+
+    where [%bc], [%ec], [%name], and [%kind] are as the above; [%filepath] contains the
+    file module path the object that the references lives in, whereas name may contain
```
Comment (JasonGross) [inline interp/dumpglob.mli:100]:
```suggestion
    file module path the object that the references lives in, whereas [%name] may contain
```

**Review by JasonGross — DISMISSED** [review]
Minor changes

File: interp/dumpglob.mli (hunk 12 above)
Comment (ejgallego) [reply interp/dumpglob.mli:52]: No easily doable with current system.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**